web/src/compose_validate.ts

```diff
@@ -795,3 +800,124 @@ export function convert_mentions_to_silent_in_direct_messages(
+    let index = content.indexOf(SPLIT_DELIMITER);
+    // skip delimiters which occur inside code blocks
+    while (index !== -1) {
+        if (!compose_ui.position_inside_code_block(content, index + 1)) {
+            return index;
+        }
+        // find next occurrence of the delimiter
+        index = content.indexOf(SPLIT_DELIMITER, index + 1);
+    }
+    return -1;
+}
+
+/**
+ * Does not trim whitespaces that appear just before text since they are
+ * markdown syntax for code blocks.
+ * @param content the string to be trimmed
+ * @returns the trimmed string
+ */
+function trim_except_whitespace_before_text(content: string): string {
+    return content.replace(/^(\n|\s*\n)+/, "").trimEnd();
+}
+
+/**
+ * Splits a message into 2 parts, if possible
+ * @param message_content the content to be split
+ * @returns a pair of strings: the first being the individual split message, the second being the
+ * remaining part of the message, which will be empty when there are no further parts to send
+ */
+export function split_message(raw_message_content: string): [string, string] {
+    if (!is_split_messages_enabled()) {
+        return [raw_message_content, ""];
+    }
+    // trim leading newlines to avoid empty messages when splitting
+    const message_content = trim_except_whitespace_before_text(raw_message_content);
+    const index = delimiter_index_outside_code_block(message_content);
+    if (index === -1) {
+        return [message_content, ""];
+    }
+    const send_content = message_content.slice(0, index);
+    // we need to trim rest_content to remove leading newlines if multiple delimiters are present
+    // in sequence, else it would lead to empty messages when `rest_content` is subsequently passed
+    // to this function for further splitting.
+    const rest_content = trim_except_whitespace_before_text(message_content.slice(index));
+    return [send_content, rest_content];
+}
+
+/**
+ * @returns the number of parts the compose content will split into
+ *
+ * mimics the message sending behaviour to count the number of message parts. We cannot directly use
+ * `message_content.split().length` because delimiters inside code blocks should be ignored.
+ */
+export function count_message_content_split_parts(): number {
+    const message_content = compose_state.message_content();
+    let count = 1;
+    let parts = split_message(message_content);
+    while (parts[1]) {
+        count += 1;
+        parts = split_message(parts[1]);
+    }
+    return count;
+}
+
+/**
+ * checks whether the message actually needs to be split, as it is possible
+ * that `split_messages` is enabled but the message doesn't contain the delimiter.
+ */
+export function will_split_into_multiple_messages(): boolean {
+    if (!is_split_messages_enabled()) {
+        return false;
+    }
+    // if there is a non-empty remaining part, splitting will occur
+    return Boolean(split_message(compose_state.message_content())[1]);
+}
```

Comment:
I think the logic functions above this point would be best put into their own little module, `compose_split_messages.ts`, which I'd expect to be just pure functions operating on strings, and that function can be written with simple node tests to illustrate the corner cases you discuss in your comments -- that'd be a much more verifiable block of code. 

`SPLIT_DELIMITER` and the associated values can live in that file too.

We can even merge that parsing logic as a prep commit, since its correctness is independent of that of the UI code to use it. I think doing that in a preparatory PR is a good next step for this project -- there's a lot of polish issues in this PR, and getting the parsing logic finished nicely will be a good checkpoint to let us shrink this.

Reply:
I have put the functions into `compose_split_messages.ts` and written tests for checking the parsing logic.